daft/execution/dynamic_construction.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import annotations
+
+from typing import Callable
+
+from daft.expressions import Expression
+from daft.logical import logical_plan
+from daft.logical.map_partition_ops import MapPartitionOp
+from daft.logical.schema import ExpressionList
+from daft.runners.partitioning import vPartition
+from daft.runners.pyrunner import LocalLogicalPartitionOpRunner
+from daft.runners.shuffle_ops import RepartitionHashOp, RepartitionRandomOp, SortOp
+
+
+class Construction:
+    """A Construction is an instruction stack + input partitions to run the instruction stack over.
+
+    Instructions can be one partition -> one partition, one->many, or many->one.
+    (To support this, instructions are typed as list[partition] -> list[partition].)
+    """
+
+    def __init__(self, inputs: list[vPartition]) -> None:
+        # Input partitions to run over.
+        self.inputs = inputs
+
+        # Instruction stack to execute.
+        self.instruction_stack: list[Callable[[list[vPartition]], list[vPartition]]] = list()
```

Comment:
Might be worth it to make this variable private and create a method to "drain" the `instruction_stack` computing the result. this would also make it easy to make it a remote function when making the ray runner.

Reply:
I think it would be better as a part of the Construction object. The way I think about the construction object is that is that it forms a composite function by stacking a set of instructions. Another option is that drain can return a function that the runner runs